python/ray/util/dask/optimizations.py

```diff
@@ -0,0 +1,168 @@
+                            MultipleReturnFunc(
+                                shuffle_group,
+                                n_parts_out,
+                            ),
+                            (self.name_input, _part_in),
+                            self.column,
+                            0,
+                            self.npartitions,
+                            self.npartitions,
+                            self.ignore_index,
+                            self.npartitions,
+                        )
+
+            return dsk
+
+    def rewrite_simple_shuffle_layer(dsk, keys):
+        if not isinstance(dsk, HighLevelGraph):
+            dsk = HighLevelGraph.from_collections(
+                id(dsk), dsk, dependencies=())
+        else:
+            dsk = dsk.copy()
+
+        layers = dsk.layers.copy()
+        for key, layer in layers.items():
+            if type(layer) is SimpleShuffleLayer:
+                dsk.layers[key] = MultipleReturnSimpleShuffleLayer.clone(layer)
+            else:
+                layer = copy.copy(layer)
+                for inner_key, inner_layer in layer.items():
+                    if type(inner_layer) is SimpleShuffleLayer:
+                        dsk.layers[key].layers[inner_key] = (
+                            MultipleReturnSimpleShuffleLayer.clone(inner_layer)
+                        )
+        return dsk
+
+    def dataframe_optimize(dsk, keys, **kwargs):
+        if not isinstance(keys, (list, set)):
+            keys = [keys]
+        keys = list(core.flatten(keys))
+
+        if not isinstance(dsk, HighLevelGraph):
+            dsk = HighLevelGraph.from_collections(
+                id(dsk), dsk, dependencies=())
+
+        dsk = rewrite_simple_shuffle_layer(dsk, keys=keys)
+        return optimize(dask, keys, **kwargs)
+else:
+
+    def dataframe_optimize(dsk, keys, **kwargs):
+        warnings.warn("Custom dataframe shuffle optimization only works on "
+                      "dask>=2020.12.0, you are on version "
+                      f"{dask.__version__}, please upgrade Dask."
+                      "Falling back to default dataframe optimizer.")
+        return optimize(dsk, keys, **kwargs)
+
+
+# Stale approaches below.
+
+
+def fuse_splits_into_multiple_return(dsk, keys):
+    if not isinstance(dsk, HighLevelGraph):
+        dsk = HighLevelGraph.from_collections(id(dsk), dsk, dependencies=())
+    else:
+        dsk = dsk.copy()
+    dependencies = dsk.dependencies.copy()
+    for k, v in dsk.items():
+        if istask(v) and v[0] == shuffle_group:
+            task_deps = dependencies[k]
+            # Only rewrite shuffle group split if all downstream dependencies
+            # are splits.
+            if all(
+                    istask(dsk[dep]) and dsk[dep][0] == operator.getitem
+                    for dep in task_deps):
+                for dep in task_deps:
+                    # Rewrite split
+                    pass
```

Comment:
I'm planning on making this more generic implementation work in the near future, so I'm slightly tempted to keep this here, despite it not doing anything.

Reply:
Is generic implementation something we discussed before, or are you thinking something else?

Reply:
It's pretty much what we discussed before!